In [1]:
import numpy as np
import pandas as pd

from sklearn.preprocessing import StandardScaler
from sklearn.model_selection import train_test_split
from sklearn.metrics import accuracy_score, confusion_matrix, classification_report

In [2]:
from sklearn.datasets import load_breast_cancer

In [3]:
data = load_breast_cancer()
X = data.data
y = data.target

In [4]:
pd.DataFrame(X).head()

0      1       2       3        4        5       6        7       8   \
0  17.99  10.38  122.80  1001.0  0.11840  0.27760  0.3001  0.14710  0.2419   
1  20.57  17.77  132.90  1326.0  0.08474  0.07864  0.0869  0.07017  0.1812   
2  19.69  21.25  130.00  1203.0  0.10960  0.15990  0.1974  0.12790  0.2069   
3  11.42  20.38   77.58   386.1  0.14250  0.28390  0.2414  0.10520  0.2597   
4  20.29  14.34  135.10  1297.0  0.10030  0.13280  0.1980  0.10430  0.1809   

        9   ...     20     21      22      23      24      25      26      27  \
0  0.07871  ...  25.38  17.33  184.60  2019.0  0.1622  0.6656  0.7119  0.2654   
1  0.05667  ...  24.99  23.41  158.80  1956.0  0.1238  0.1866  0.2416  0.1860   
2  0.05999  ...  23.57  25.53  152.50  1709.0  0.1444  0.4245  0.4504  0.2430   
3  0.09744  ...  14.91  26.50   98.87   567.7  0.2098  0.8663  0.6869  0.2575   
4  0.05883  ...  22.54  16.67  152.20  1575.0  0.1374  0.2050  0.4000  0.1625   

       28       29  
0  0.4601  0.11890  
1  0.2750  0.08902  
2  0.3613  0.08758  
3  0.6638  0.17300  
4  0.2364  0.07678  

[5 rows x 30 columns]

In [5]:
X.shape

(569, 30)

In [6]:
# test_training_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=42,stratify=y)

#feature scaling
scaler=StandardScaler()
X_train=scaler.fit_transform(X_train)
X_test=scaler.transform(X_test)

In [7]:
from sklearn.neighbors import KNeighborsClassifier
from sklearn.linear_model import LogisticRegression, LinearRegression
from sklearn.tree import DecisionTreeClassifier
from sklearn.ensemble import RandomForestClassifier, BaggingClassifier
from sklearn.svm import SVC
from sklearn.naive_bayes import GaussianNB
from sklearn.metrics import accuracy_score, precision_score, recall_score, f1_score, confusion_matrix

In [8]:
# empty list to store the result of models
models_results = []

In [9]:
#1. KNN
#train the model
knn_model = KNeighborsClassifier()
knn_model.fit(X_train, y_train)
#make predictions
y_pred_knn = knn_model.predict(X_test)

In [10]:
#calculating metrics
knn_acc = accuracy_score(y_test, y_pred_knn)
knn_prec = precision_score(y_test, y_pred_knn)
knn_rec = recall_score(y_test, y_pred_knn)
knn_f1 = f1_score(y_test, y_pred_knn)
knn_cm = confusion_matrix(y_test, y_pred_knn)

In [11]:
print("----KNN Evaluation Metrics----")
print(f"Accuracy: {knn_acc:.4f}")
print(f"Precision: {knn_prec:.4f}")
print(f"Recall: {knn_rec:.4f}")
print(f"F1 Score: {knn_f1:.4f}")
print("Confusion Matrix:\n", knn_cm)

----KNN Evaluation Metrics----
Accuracy: 0.9591
Precision: 0.9386
Recall: 1.0000
F1 Score: 0.9683
Confusion Matrix:
 [[ 57   7]
 [  0 107]]


In [12]:
#saving the metrics in empty list
models_results.append({
    "Model": "KNN",
    "Accuracy": knn_acc,
    "Precision": knn_prec,
    "Recall": knn_rec,
    "F1 Score": knn_f1,
    "Confusion Matrix": knn_cm
})

In [13]:
#2. Logistic Regression
logistic_model = LogisticRegression()
logistic_model.fit(X_train, y_train)
y_pred_logistic = logistic_model.predict(X_test)

In [14]:
logistic_acc = accuracy_score(y_test, y_pred_logistic)
logistic_prec = precision_score(y_test, y_pred_logistic)
logistic_rec = recall_score(y_test, y_pred_logistic)
logistic_f1 = f1_score(y_test, y_pred_logistic)
logistic_cm = confusion_matrix(y_test, y_pred_logistic)

In [15]:
print("----Logistic Evaluation Matrix----")
print("Accuracy: ",logistic_acc)
print(f"Accuracy: {logistic_acc:.4f}")
print(f"Precision: {logistic_prec:.4f}")
print(f"Recall: {logistic_rec:.4f}")
print(f"F1 Score: {logistic_f1:.4f}")
print("Confusion Matrix:\n",logistic_cm)

----Logistic Evaluation Matrix----
Accuracy:  0.9883040935672515
Accuracy: 0.9883
Precision: 0.9907
Recall: 0.9907
F1 Score: 0.9907
Confusion Matrix:
 [[ 63   1]
 [  1 106]]


In [16]:
models_results.append({
    "Model": "Logistic Regression",
    "Accuracy": logistic_acc,
    "Precision": logistic_prec,
    "Recall": logistic_rec,
    "F1 Score": logistic_f1,
    "Confusion Matrix": logistic_cm
})

In [17]:
#3. L1 Regularization
l1_model = LogisticRegression(
    penalty='l1',
    solver='liblinear',
    C=1.0
)
l1_model.fit(X_train, y_train)
y_pred_l1 = l1_model.predict(X_test)

In [18]:
l1_acc = accuracy_score(y_test, y_pred_l1)
l1_prec = precision_score(y_test, y_pred_l1)
l1_rec = recall_score(y_test, y_pred_l1)
l1_f1 = f1_score(y_test, y_pred_l1)
l1_cm = confusion_matrix(y_test, y_pred_l1)

In [19]:
print("----L1 Regularization Evaluation Matrix----")
print(f"Accuracy: {l1_acc:.4f}")
print(f"Precision: {l1_prec:.4f}")
print(f"Recall: {l1_rec:.4f}")
print(f"F1 Score: {l1_f1:.4f}")
print("Confusion Matrix:\n",l1_cm)

----L1 Regularization Evaluation Matrix----
Accuracy: 0.9708
Precision: 0.9722
Recall: 0.9813
F1 Score: 0.9767
Confusion Matrix:
 [[ 61   3]
 [  2 105]]


In [20]:
models_results.append({
    "Model": "L1 Regularization",
    "Accuracy": l1_acc,
    "Precision": l1_prec,
    "Recall": l1_rec,
    "F1 Score": l1_f1,
    "Confusion Matrix": l1_cm
})

In [22]:
#4. L2 Regularization
l2_model = LogisticRegression(
    penalty = 'l2',
    solver = 'lbfgs',
    C = 1.0
)
l2_model.fit(X_train, y_train)
y_pred_l2 = l2_model.predict(X_test)

In [23]:
l2_acc = accuracy_score(y_test, y_pred_l2)
l2_prec = precision_score(y_test, y_pred_l2)
l2_rec = recall_score(y_test, y_pred_l2)
l2_f1 = f1_score(y_test, y_pred_l2)
l2_cm = confusion_matrix(y_test, y_pred_l2)

In [25]:
print("----L2 Regularization Evaluation Matrix----")
print(f"Accuracy: {l2_acc:.4f}")
print(f"Precision: {l2_prec:.4f}")
print(f"Recall: {l2_rec:.4f}")
print(f"F1 Score: {l2_f1:.4f}")
print("Confusion Matrix:\n",l2_cm)

----L2 Regularization Evaluation Matrix----
Accuracy: 0.9883
Precision: 0.9907
Recall: 0.9907
F1 Score: 0.9907
Confusion Matrix:
 [[ 63   1]
 [  1 106]]


In [26]:
models_results.append({
    "Model": "L2 Regularization",
    "Accuracy": l2_acc,
    "Precision": l2_prec,
    "Recall": l2_rec,
    "F1 Score": l2_f1,
    "Confusion matrix": l2_cm
})

In [30]:
#5. Linear Regression
linear_model = LinearRegression()
linear_model.fit(X_train, y_train)
y_pred_linear = linear_model.predict(X_test)
y_pred_linear = (y_pred_linear > 0.5).astype(int) #to convert continuous predictions to binary classes using a 0.5 threshold

In [31]:
linear_acc = accuracy_score(y_test, y_pred_linear)
linear_prec = precision_score(y_test, y_pred_linear)
linear_rec = recall_score(y_test, y_pred_linear)
linear_f1 = f1_score(y_test, y_pred_linear)
linear_cm = confusion_matrix(y_test, y_pred_linear)

In [32]:
print("----Linear Regression Evaluation Matrix----")
print(f"Accuracy: {linear_acc:.4f}")
print(f"Precision: {linear_prec:.4f}")
print(f"Recall: {linear_rec:.4f}")
print(f"F1 Score: {linear_f1:.4f}")
print("Confusion Matrix:\n",linear_cm)

----Linear Regression Evaluation Matrix----
Accuracy: 0.9298
Precision: 0.9060
Recall: 0.9907
F1 Score: 0.9464
Confusion Matrix:
 [[ 53  11]
 [  1 106]]


In [33]:
models_results.append({
    "Model": "Linear Regression",
    "Accuracy": linear_acc,
    "Precision": linear_prec,
    "Recall": linear_rec,
    "F1 Score": linear_f1,
    "Confusion Matrix": linear_cm
})

In [34]:
#6. Decision Tree
dt_model = DecisionTreeClassifier()
dt_model.fit(X_train, y_train)
y_pred_dt = dt_model.predict(X_test)

In [35]:
dt_acc = accuracy_score(y_test, y_pred_dt)
dt_prec = precision_score(y_test, y_pred_dt)
dt_rec = recall_score(y_test, y_pred_dt)
dt_f1 = f1_score(y_test, y_pred_dt)
dt_cm = confusion_matrix(y_test, y_pred_dt)

In [36]:
print("----Decision Tree Evaluation Matrix----")
print(f"Accuracy: {dt_acc:.4f}")
print(f"Precision: {dt_prec:.4f}")
print(f"Recall: {dt_rec:.4f}")
print(f"F1 Score: {dt_f1:.4f}")
print("Confusion Matrix:\n",dt_cm)

----Decision Tree Evaluation Matrix----
Accuracy: 0.9240
Precision: 0.9434
Recall: 0.9346
F1 Score: 0.9390
Confusion Matrix:
 [[ 58   6]
 [  7 100]]


In [37]:
models_results.append({
    "Model": "Decision Tree",
    "Accuracy": dt_acc,
    "Precision": dt_prec,
    "Recall": dt_rec,
    "F1 Score": dt_f1,
    "Confusion Matrix": dt_cm
})

In [47]:

#7. Bagging(Decision Tree Base)
bagging_model = BaggingClassifier(
    estimator=DecisionTreeClassifier(),n_estimators=100,random_state=42
    )
bagging_model.fit(X_train, y_train)
y_pred_bagging = bagging_model.predict(X_test)

In [48]:
bagging_acc = accuracy_score(y_test, y_pred_bagging)
bagging_prec = precision_score(y_test, y_pred_bagging)
bagging_rec = recall_score(y_test, y_pred_bagging)
bagging_f1 = f1_score(y_test, y_pred_bagging)
bagging_cm = confusion_matrix(y_test, y_pred_bagging)

In [49]:
print("----Bagging Classifier Evaluation Matrix----")
print(f"Accuracy: {bagging_acc:.4f}")
print(f"Precision: {bagging_prec:.4f}")
print(f"Recall: {bagging_rec:.4f}")
print(f"F1 Score: {bagging_f1:.4f}")
print("Confusion Matrix:\n",bagging_cm)

----Bagging Classifier Evaluation Matrix----
Accuracy: 0.9474
Precision: 0.9455
Recall: 0.9720
F1 Score: 0.9585
Confusion Matrix:
 [[ 58   6]
 [  3 104]]


In [50]:
models_results.append({
    "Model": "Bagging Classifier",
    "Accuracy": bagging_acc,
    "Precision": bagging_prec,
    "Recall": bagging_rec,
    "F1 Score": bagging_f1,
    "Confusion Matrix": bagging_cm
})

In [52]:
#8. Random Forest
rf_model = RandomForestClassifier()
rf_model.fit(X_train, y_train)
y_pred_rf = rf_model.predict(X_test)

In [53]:
rf_acc = accuracy_score(y_test, y_pred_rf)
rf_prec = precision_score(y_test, y_pred_rf)
rf_rec = recall_score(y_test, y_pred_rf)
rf_f1 = f1_score(y_test, y_pred_rf)
rf_cm = confusion_matrix(y_test, y_pred_rf)

In [54]:
print("----Random Forest Evaluation Matrix----")
print(f"Accuracy: {rf_acc:.4f}")
print(f"Precision: {rf_prec:.4f}")
print(f"Recall: {rf_rec:.4f}")
print(f"F1  score: {rf_f1:.4f}")
print("Confusion matrix:\n",rf_cm)

----Random Forest Evaluation Matrix----
Accuracy: 0.9474
Precision: 0.9455
Recall: 0.9720
F1  score: 0.9585
Confusion matrix:
 [[ 58   6]
 [  3 104]]


In [55]:
models_results.append({
    "Model": "Random Forest",
    "Accuracy": rf_acc,
    "Precision": rf_prec,
    "Recall": rf_rec,
    "F1 Score": rf_f1,
    "Confusion Matrix": rf_cm
})

In [63]:
#9. Support Vector Machine
svm_model_lnr = SVC(kernel='linear', C=100)
svm_model_lnr.fit(X_train, y_train)
y_pred_svm_lnr = svm_model.predict(X_test)

In [64]:
svm_lnr_acc = accuracy_score(y_test, y_pred_svm_lnr)
svm_lnr_prec = precision_score(y_test, y_pred_svm_lnr)
svm_lnr_rec = recall_score(y_test, y_pred_svm_lnr)
svm_lnr_f1 = f1_score(y_test, y_pred_svm_lnr)
svm_lnr_cm = confusion_matrix(y_test, y_pred_svm_lnr)

In [65]:
print("----SVM Linear Evaluation Matrix----")
print(f"Accuracy: {svm_lnr_acc:.4f}")
print(f"Precision: {svm_lnr_prec:.4f}")
print(f"Recall: {svm_lnr_rec:.4f}")
print(f"F1 Score: {svm_lnr_f1:.4f}")
print("Confusion Matrix:\n",svm_lnr_cm)

----SVM Linear Evaluation Matrix----
Accuracy: 0.9825
Precision: 0.9815
Recall: 0.9907
F1 Score: 0.9860
Confusion Matrix:
 [[ 62   2]
 [  1 106]]


In [66]:
models_results.append({
    "Model": "SVM Linear",
    "Accuracy": svm_lnr_acc,
    "Precision": svm_lnr_prec,
    "Recall": svm_lnr_rec,
    "F1 Score": svm_lnr_f1,
    "Confusion Matrix": svm_lnr_cm
})

In [67]:
svm_model_rbf = SVC(kernel='rbf', C=100, gamma='scale')
svm_model_rbf.fit(X_train, y_train)
y_pred_svm_rbf = svm_model_rbf.predict(X_test)

In [68]:
svm_rbf_acc = accuracy_score(y_test, y_pred_svm_rbf)
svm_rbf_prec = precision_score(y_test, y_pred_svm_rbf)
svm_rbf_rec = recall_score(y_test, y_pred_svm_rbf)
svm_rbf_f1 = f1_score(y_test, y_pred_svm_rbf)
svm_rbf_cm = confusion_matrix(y_test, y_pred_svm_rbf)

In [69]:
print("----SVM RBF Evaluation Matrix----")
print(f"Accuracy: {svm_rbf_acc:.4f}")
print(f"Precision: {svm_rbf_prec:.4f}")
print(f"Recall: {svm_rbf_rec:.4f}")
print(f"F1 Score: {svm_rbf_f1:.4f}")
print("Confusion Matrix:\n",svm_rbf_cm)

----SVM RBF Evaluation Matrix----
Accuracy: 0.9240
Precision: 0.9352
Recall: 0.9439
F1 Score: 0.9395
Confusion Matrix:
 [[ 57   7]
 [  6 101]]


In [70]:
models_results.append({
    "Model": "SVM RBF",
    "Accuracy": svm_rbf_acc,
    "Precision": svm_rbf_prec,
    "Recall": svm_rbf_rec,
    "F1 Score": svm_rbf_f1,
    "Confusion matrix": svm_rbf_cm
})

In [72]:
#10. Linear Discriminant Analysis (LDA)
from sklearn.discriminant_analysis import LinearDiscriminantAnalysis
lda_model = LinearDiscriminantAnalysis()
lda_model.fit(X_train, y_train)
y_pred_lda = lda_model.predict(X_test)

In [73]:
lda_acc = accuracy_score(y_test, y_pred_lda)
lda_prec = precision_score(y_test, y_pred_lda)
lda_rec = recall_score(y_test, y_pred_lda)
lda_f1 = f1_score(y_test, y_pred_lda)
lda_cm = confusion_matrix(y_test, y_pred_lda)

In [74]:
print("----LDA Evaluation Matrix----")
print(f"Accuracy: {lda_acc:.4f}")
print(f"Precision: {lda_prec:.4f}")
print(f"Recall: {lda_rec:.4f}")
print(f"F1 Score: {lda_f1:.4f}")
print("Confusion Matrix:\n", lda_cm)

----LDA Evaluation Matrix----
Accuracy: 0.9357
Precision: 0.9138
Recall: 0.9907
F1 Score: 0.9507
Confusion Matrix:
 [[ 54  10]
 [  1 106]]


In [75]:
models_results.append({
    "Model": "LDA",
    "Accuracy": lda_acc,
    "Precision": lda_prec,
    "Recall": lda_rec,
    "F1 Score": lda_f1,
    "Confusion Matrix": lda_cm
})

In [76]:
#11. Naive Bayes
nb_model = GaussianNB()
nb_model.fit(X_train, y_train)
y_pred_nb = nb_model.predict(X_test)

In [78]:
nb_acc = accuracy_score(y_test, y_pred_nb)
nb_prec = precision_score(y_test, y_pred_nb)
nb_rec = recall_score(y_test, y_pred_nb)
nb_f1 = f1_score(y_test, y_pred_nb)
nb_cm = confusion_matrix(y_test, y_pred_nb)

In [79]:
print("----Naive Bayes Evaluation Matrix----")
print(f"Accuracy: {nb_acc:.4f}")
print(f"Precision: {nb_prec:.4f}")
print(f"Recall: {nb_rec:.4f}")
print(f"F1 Score: {nb_f1:.4f}")
print("Confusion Matrix:\n",nb_cm)

----Naive Bayes Evaluation Matrix----
Accuracy: 0.9357
Precision: 0.9364
Recall: 0.9626
F1 Score: 0.9493
Confusion Matrix:
 [[ 57   7]
 [  4 103]]


In [80]:
models_results.append({
    "Model": "Naive Bayes",
    "Accuracy": nb_acc,
    "Precision": nb_prec,
    "Recall": nb_rec,
    "F1 Score": nb_f1,
    "Confusion Matrix": nb_cm
})

In [82]:
#Comparison Table for all models
final_table = pd.DataFrame(models_results)
final_table

Model  Accuracy  Precision    Recall  F1 Score  \
0                   KNN  0.959064   0.938596  1.000000  0.968326   
1   Logistic Regression  0.988304   0.990654  0.990654  0.990654   
2     L1 Regularization  0.970760   0.972222  0.981308  0.976744   
3     L2 Regularization  0.988304   0.990654  0.990654  0.990654   
4     Linear Regression  0.929825   0.905983  0.990654  0.946429   
5         Decision Tree  0.923977   0.943396  0.934579  0.938967   
6    Bagging Classifier  0.947368   0.945455  0.971963  0.958525   
7    Bagging Classifier  0.947368   0.945455  0.971963  0.958525   
8    Bagging Classifier  0.947368   0.945455  0.971963  0.958525   
9         Random Forest  0.947368   0.945455  0.971963  0.958525   
10           SVM Linear  0.982456   0.981481  0.990654  0.986047   
11           SVM Linear  0.982456   0.981481  0.990654  0.986047   
12              SVM RBF  0.923977   0.935185  0.943925  0.939535   
13                  LDA  0.935673   0.913793  0.990654  0.950673   
14          Naive Bayes  0.935673   0.936364  0.962617  0.949309   

        Confusion Matrix     Confusion matrix  
0    [[57, 7], [0, 107]]                  NaN  
1    [[63, 1], [1, 106]]                  NaN  
2    [[61, 3], [2, 105]]                  NaN  
3                    NaN  [[63, 1], [1, 106]]  
4   [[53, 11], [1, 106]]                  NaN  
5    [[58, 6], [7, 100]]                  NaN  
6    [[58, 6], [3, 104]]                  NaN  
7    [[58, 6], [3, 104]]                  NaN  
8    [[58, 6], [3, 104]]                  NaN  
9    [[58, 6], [3, 104]]                  NaN  
10   [[62, 2], [1, 106]]                  NaN  
11   [[62, 2], [1, 106]]                  NaN  
12                   NaN  [[57, 7], [6, 101]]  
13  [[54, 10], [1, 106]]                  NaN  
14   [[57, 7], [4, 103]]                  NaN

**1. Analyze Bias-Variance Behavior**
When comparing the base Decision Tree to the ensemble models, a clear shift in the bias-variance tradeoff is visible. The single Decision Tree achieved ann  accuracy of ~92.4% and a recall of ~93.4%. However, ensemble methods like Random Forest and Bagging both improved performance, jumping to ~94.7% accuracy and ~97.2% recall. This indicates that the single Decision Tree likely suffered from high variance (overfitting the training data). By bootstrapping the data and aggregating the results of multiple trees, the Bagging and Random Forest classifiers successfully reduced that variance, leading to a model that generalizes much better on unseen test data.

**2. Comparing L1 vs L2 Regularization Effects**
When evaluating the Logistic Regression models, L2 regularization (Ridge) outperformed L1 regularization (Lasso). The L2 model achieved an accuracy of ~98.8% and a recall of ~99.0%, perfectly matching the unpenalized Logistic Regression model. In contrast, the L1 model's accuracy dropped slightly to ~97.0%. Because L1 regularization applies an absolute value penalty that pushes leess important feature weights to exactly zero (performing aggressive feature selection), it likely discarded some minor but useful information among the 30 numerical features. L2 regularization, which squre the weights and shrinks them without eliminating them entirely, proved much more effective for capturing the undrlying patterns in this specific medical dataset.

**3. Recommending the Best Model with Proper Reasoning**
For this project, the primary objective is medical risk prediction. In a healthcare context, the cost of a False Negative (predicting a patient is 'Benign' when they actually have a 'Malignant' high-risk condition) is catastrophic. Therefore, while *Accuracy* is important, *Recall* is the most critical evaluation metric.

Based onn the results, I recommend the *Logistic Regression (with L2 Regularization)* as the optimal model. It provides an exceptional balance, delivering the highest overall accuracy (~98.8%) while maintaining a stellar recall of ~99.0%.

On the other hand, if the strict medical requirement is to have absolutely zero False Negatives, the *K-Nearest Neighbors (KNN)* model could be considered, as it achieved a perfect Recall of 1.0 (100%), though it sacrifices a small amount of overall accuracy (~95.9%) to achieve this.

